atlasdb-cassandra/src/test/java/com/palantir/atlasdb/cassandra/ReloadingCloseableContainerTest.java

```diff
@@ -0,0 +1,325 @@
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.palantir.atlasdb.cassandra;
+
+import static org.assertj.core.api.AssertionsForClassTypes.assertThat;
+import static org.assertj.core.api.AssertionsForClassTypes.assertThatThrownBy;
+import static org.mockito.ArgumentMatchers.any;
+import static org.mockito.Mockito.doThrow;
+import static org.mockito.Mockito.times;
+import static org.mockito.Mockito.verify;
+import static org.mockito.Mockito.verifyNoInteractions;
+import static org.mockito.Mockito.when;
+
+import com.palantir.atlasdb.keyvalue.cassandra.ReloadingCloseableContainer;
+import com.palantir.common.concurrent.PTExecutors;
+import com.palantir.logsafe.exceptions.SafeIllegalStateException;
+import com.palantir.refreshable.Refreshable;
+import com.palantir.refreshable.SettableRefreshable;
+import java.io.Closeable;
+import java.io.IOException;
+import java.time.Duration;
+import java.util.concurrent.BrokenBarrierException;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.CyclicBarrier;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Future;
+import java.util.concurrent.ThreadLocalRandom;
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.function.Function;
+import java.util.stream.IntStream;
+import org.awaitility.Awaitility;
+import org.junit.Before;
+import org.junit.Test;
+import org.junit.runner.RunWith;
+import org.mockito.Mock;
+import org.mockito.junit.MockitoJUnitRunner;
+
+@RunWith(MockitoJUnitRunner.class)
+public class ReloadingCloseableContainerTest {
+    private static final ExecutorService executor =
+            PTExecutors.newCachedThreadPool(ReloadingCloseableContainerTest.class.getName());
+
+    private static final int INITIAL_VALUE = 0;
+    private static final int UPDATED_VALUE = 1;
+
+    @Mock
+    private Closeable mockCloseable;
+
+    @Mock
+    private Closeable mockCloseableExtra;
+
+    @Mock
+    private Function<Integer, Closeable> closeableFactory;
+
+    private SettableRefreshable<Integer> settableRefreshable;
+
+    @Before
+    public void setUp() {
+        settableRefreshable = Refreshable.create(INITIAL_VALUE);
+        when(closeableFactory.apply(INITIAL_VALUE)).thenReturn(mockCloseable);
+        when(closeableFactory.apply(UPDATED_VALUE)).thenReturn(mockCloseableExtra);
+    }
+
+    @Test
+    public void closeBlocksDuringRunWithInvocation() throws BrokenBarrierException, InterruptedException {
```

Comment:
Let me clarify: this test is trying to:

1. run a task, and wait
2. close, and show that closing blocks until the task completes
3. finish the task
4. show that close happens
?

I think the way to do this more easily is with `CountDownLatch`:
```java
    @Test
    public void closeBlocksDuringRunWithInvocation() throws IOException {
        ReloadingCloseableContainer<Closeable> container = getContainer();
        CountDownLatch taskRunning = new CountDownLatch(1);
        CountDownLatch closeRunning = new CountDownLatch(1);
        CountDownLatch readyToClose = new CountDownLatch(1);

        executor.execute(() -> container.executeWithResource(resource -> {
            taskRunning.countDown();
            awaitLatch(readyToClose);
            verifyNoInteractions(resource);
        }));

        executor.execute(() -> {
            closeRunning.countDown();
            awaitLatch(taskRunning);
            container.close();
        });
        
        awaitLatch(taskRunning);
        awaitLatch(closeRunning);
        assertThat(container.isClosed()).isFalse();

        verifyNoInteractions(mockCloseable);
        readyToClose.countDown();
        
        Awaitility.await().atMost(Duration.ofSeconds(1)).until(container::isClosed);
        verify(mockCloseable).close();
    }
```